### PR TITLE
Fuzzy Search

### LINKED ISSUE #14337: Global search should be fuzzy
I'm not sure exactly how far we should go with this, but currently our global search is pretty rigid.

One precise example is that you can't omit a word in a multi-word item you're searching for. So if I'm looking for "Anonymous Usage Statistics" but can't remember what it's called _exactly_ and you type "anonymous statistics", you won't get any results, which feels broken:

fuzzy-search

You also have to get the _order_ right — you won't get results if you type "statistics usage anonymous."

We also don't seem to do partial or similar string search, so in my example above, if i searched for "anon usage stats" I would also not get results.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by tsmacdonald**
[Fixes #14337]

Enlarges the search results by searching for each of the search tokens separately, then sorts them with several heuristics.

The heuristics seem ~fine on my machine, but the real test will be on stats or somewhere else with a much larger dataset. There's no One Right Answer, just our perception of whether things seem to be in the right order.

If they are, then great.

If they're not, the more sophisticated (industry-standard) approach is [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf), which has the downside of needing to maintain a frequency list of how many occurrences of a given term we have. I think we could do that no problem, but it is another table / more storage required / engineering effort to make it stay in sync. If we did have such a table it would pave the way for "did you mean"-style spelling suggestions.

### CODE REVIEW

**Comment by jeff303** [comment]
> [Fixes #14337]
> 
> Enlarges the search results by searching for each of the search tokens separately, then sorts them with several heuristics.
> 
> The heuristics seem ~fine on my machine, but the real test will be on stats or somewhere else with a much larger dataset. There's no One Right Answer, just our perception of whether things seem to be in the right order.
> 
> If they are, then great.
> 
> If they're not, the more sophisticated (industry-standard) approach is [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf), which has the downside of needing to maintain a frequency list of how many occurrences of a given term we have. I think we could do that no problem, but it is another table / more storage required / engineering effort to make it stay in sync. If we did have such a table it would pave the way for "did you mean"-style spelling suggestions.

I think at a certain point, it would probably make sense to use something like Lucene, right?  ElasticSearch is even nicer, although I doubt we really want to open that can of worms since it adds a lot of infrastructure complexity.  But it can also handle things like stemming, language-aware tokenization, etc.

**Comment by dpsutton** [comment]
I've been through this kind of thing at the last job and it gets pretty maddening. One problem I remember is when looking for compound terms that you know are correct, like a name "Bob Anderson". If you get a bunch of bobs (not andersons) and then a bunch of andersons (not necessarily bobs) and then at the bottom (or perhaps not at all) you get the bob anderson you're actually looking for. Tokenizing too much and discarding the full query can lead to this. 

I'm looking at your implementation of `search-string-clause` and it looks like the terms are broken up and never considered as a whole (unless I'm misreading it). Is it possible to add the full clause back to each of the searched columns as well? A good sample use case from the original ticket would be searching for "Anonymous Usage" which should match "Anonymous Usage Statistics" presumably first before all "anonymous" things, then all "Usage" things

**Comment by jeff303** [comment]
> I'm looking at your implementation of `search-string-clause` and it looks like the terms are broken up and never considered as a whole (unless I'm misreading it). Is it possible to add the full clause back to each of the searched columns as well? A good sample use case from the original ticket would be searching for "Anonymous Usage" which should match "Anonymous Usage Statistics" presumably first before all "anonymous" things, then all "Usage" things

Yeah, an n-gram hit should ideally rank higher than a single term hit.

File: src/metabase/api/search.clj [hunk 1]
```diff
@@ -386,16 +379,18 @@
...
-      (for [row results
-            :when (check-permissions-for-model row)]
-        ;; MySQL returns `:favorite` and `:archived` as `1` or `0` so convert those to boolean as needed
-        (-> row
-            (update :favorite bit->boolean)
-            (update :archived bit->boolean))))))
+          results      (db/query search-query :max-rows search-max-results)]
+      ;; sort by [score model name]
+      (sort-by (juxt
+                (comp - (partial search/score (:search-string search-ctx)))
```
Comment (dpsutton) [inline src/metabase/api/search.clj:385]: the pattern `(sort-by - ...)` has some pitfalls with floats: https://clojure.org/guides/comparators#_beware_using_subtraction_in_a_comparator

tldr the output of compare must be a integer. if you return a float, it is cast and this can lead to loss of resolution:
```
(sort - [1.0 0.9 0.8 0.7]) #_ -> (1.0 0.9 0.8 0.7)
(sort - [1 2 3 4]) #_ -> (1 2 3 4)
```

**Comment by tsmacdonald** [comment]
>Yeah, an n-gram hit should ideally rank higher than a single term hit.

That's the `consecutivity-scorer`

File: src/metabase/api/search.clj (hunk 1 above)
Comment (tsmacdonald) [reply src/metabase/api/search.clj:385]: TIL, thanks!

Unless I'm misunderstanding, using the unary `-` as the `keyfn` will invoke `compare`, which uses traditional `<`, `==`, and `>` to arrive at `-1`, `0`, and `1`. This code uses rationals, but things seem good for both rationals and floats:

```
(sort-by - [1/2 1/3 2/3 4/3 10/3])
;; => (10/3 4/3 2/3 1/2 1/3)
(sort-by - [0.1 0.2 0.4 0.3 0.9 0.15 1.5])
;; => (1.5 0.9 0.4 0.3 0.2 0.15 0.1)
```


```
(defn sort-by
  "Returns a sorted sequence of the items in coll, where the sort
  order is determined by comparing (keyfn item).  If no comparator is
  supplied, uses compare.  comparator must implement
  java.util.Comparator.  Guaranteed to be stable: equal elements will
  not be reordered.  If coll is a Java array, it will be modified.  To
  avoid this, sort a copy of the array."
  {:added "1.0"
   :static true}
  ([keyfn coll]
   (sort-by keyfn compare coll))
  ([keyfn ^java.util.Comparator comp coll]
   (sort (fn [x y] (. comp (compare (keyfn x) (keyfn y)))) coll)))
   ```

File: src/metabase/search.clj [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  (str/lower-case query))
+
+(s/defn tokenize :- [s/Str]
+  "Break a search `query` into its constituent tokens"
+  [query :- s/Str]
+  (filter seq
+          (str/split query #"\s+")))
+
+(def ^:private largest-common-subseq-length
+  (memoize/fifo
```
Comment (dpsutton) [inline src/metabase/search.clj:20]: the `fifo` here defaults to a queue size of 32. That seems pretty small and perhaps could even run out of useful cached results in the recursions here. I think you can do `(memoize/fif (fn ...) :fifo/threshold 1000)` or perhaps even higher. This seems like useful to keep a cache of these things and fundamentally the data are just `{string -> integer}` so we aren't holding on to a huge amount in the cache.

File: src/metabase/api/search.clj (hunk 1 above)
Comment (dpsutton) [reply src/metabase/api/search.clj:385]: oh right. good point about the compare versus keyfn.

**Comment by dpsutton** [comment]
i'm doing the following:

```
  (search/score "anonymous usage" (result-row "anonymous usage statistics"))
  ;; => 3.5
  (search/score "Anonymous Usage" (result-row "Anonymous Usage Statistics"))
  ;; => 0.0
```
Am I violating some expectation here? Is the search term supposed to have been downcased or normalized somehow before `metabase.search/score`?

**Comment by dpsutton** [comment]
I think some kind of test like the following would help piecing this together as well:
```clojure
  (sort-by (comp - second)
           (map (fn [db-row] [db-row (search/score "anonymous statistics" (result-row db-row))])
                ["Anonymous Usage Statistics"
                 "Anonymous Users"
                 "Results from Anonymous"
                 "Car Statistics"
                 "Statistics from Anonymous People"
                 "Anonymous"
                 "Statistics"
                 "Statistics Anonymous"]))
;; => (["Anonymous Usage Statistics" 3.0] ["Statistics from Anonymous People" 3.0] ["Statistics Anonymous" 3.0] ["Anonymous Users" 1.75] ["Results from Anonymous" 1.75] ["Car Statistics" 1.75] ["Anonymous" 1.75] ["Statistics" 1.75])
```
It demonstrates how all the scorer's piece together. Not required but i didn't see the n-gram highlighting at first glance. 

Implementation looks really clean. i like it. Nice work!

File: src/metabase/search.clj [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+(defmethod searchable-columns-for-model :default
+  [_]
+  [:name])
+
+(defmethod searchable-columns-for-model (class Table)
+  [_]
+  [:name
+   :display_name])
+
```
Comment (jeff303) [inline src/metabase/search.clj:58]: What about `Field` as well?  I believe it has `:name` and `:display_name`?

Comment (tsmacdonald) [reply src/metabase/search.clj:58]: We don't search `Field` s at all right now. There's [an open issue](https://github.com/metabase/metabase/issues/7019), but before we tackle it we need to figure out what clicking a `Field` result should even take you to (especially if you're not an admin).

Kyle is brainstorming on this—the use cases for searching for `Field` s may well be solved by some other feature

**Comment by tsmacdonald** [comment]
> Am I violating some expectation here? Is the search term supposed to have been downcased or normalized somehow before metabase.search/score?

d'oh, fixed
